### DISCUSSION TITLE
Ollama Support

### DISCUSSION THREAD

**Discussion by rhyswynn** [dfinke/PSAI#3] [Ideas]
Hi there! Great contribution to the community, as always, thank you!

Any thoughts on adding Ollama support? Ollama allows for integration with many other open source local LLMs, which can be valuable tools for developing LLM queries without spending on tokens. I have integrated Ollama support with a different Azure/OpenAI LLM query framework in Python, so it can definitely be done without too much effort.  Thanks again!

**Comment by dfinke** [comment]
Thank you @rhyswynn. Definitely want to have a `plugin` capability for different LLMs.

First, PSAI needs to be on parity with the OpenAI Python SDK. They just dropped v2 so that is my current focus.

The challenge for integration is that PSAI is using the new Assistant API which is a conversational stateless approach. The others are completion and stateful. I have not come up with a good way to build this.

The current frameworks do not use the stateless OpenAI APIs. They use the previous version. Plus Ollama, Gemini, Claude use the stateful older approach.

https://gist.github.com/dfinke/9e5c10e8ad1333af1e4e426dd3dbb5e3 - this is simple psllm.ps1 that works with Gemini, Claude and OpenAI older approach.

Having a `plugin` LLM capabilty enables really interesting and useful scenarios.

Right now I am on the hook to do a book for Manning "AI Assisted PowerShell" so that is a priority.

Happy to discuss and review PRs that could work.

**Comment by dfinke** [comment]
@rhyswynn What kind of machine are you running the OLlama on?

**Reply by rhyswynn** [reply]
A few year old workstation, 24GB RTX GPU. I read an article on running the 70B model with less GPU memory, but that wasn't using Ollama.  The docs say you need 16GB for a 7B model. https://github.com/ollama/ollama#model-library

**Comment by ABrauser** [comment]
Hi all,

I just managed to run Ollama on my desktop via Docker-Desktop. (Unfortunately, it is only CPU-based because I have an AMD graphics card, and there is no support for it yet. However, for my use cases, the CPU R9 3900x with 32GB RAM is decent enough.) I am also very interested in integrating a local Docker-based LLM. I'm using "open-webui
," which has an API (http://localhost:11434/) and API keys that can be generated on the local website. How much effort would it be to add this provider, similar to the AzureOpenAI approach?

Currently, I am experimenting with the PSAI module and am highly impressed with the concept of offloading decision-making tasks within PowerShell scripts to an AI. (@dfinke  Thank you very much for your work so far!) My specific goal is to automate the process of renaming and organizing documents that I have scanned as PDFs (with OCR) locally on my desktop using PowerShell. 

Any advice or guidance on this would be greatly appreciated. Thank you!

**Comment by dfinke** [comment]
Thank you @ABrauser. I haven't looked at the Ollama implementation. Don't think it would be hard. Like I said in the prev comment, I am thinking through a `plugin` architecture. 

@ABrauser @rhyswynn as a note, GitHub is previewing access to ~all the models. I have a private working copy. It only works with the OpenAI models. Other models have different request/response approaches. GitHub has abstracted that and enable `passthru` for specific model parameters.

@ABrauser as a workaround, it probably won't work after the plugins are implemented. Checkout `Set-OAIProvider` and `Invoke-OAIBeta` functions. You might be able to expand it for Ollama support.

Let me know how it goes or have questions.

**Comment by dfinke** [comment]
update https://github.com/dfinke/PSAI/pull/48
